compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/CfaUtils.kt

```diff
@@ -120,10 +127,156 @@ class PropertyInitializationInfoCollector(private val localProperties: Set<FirPr
+        return if (reset)
+            overwriteRange(dataForNode, symbol, EventOccurrencesRange.ZERO, ::PathAwarePropertyInitializationInfo)
+        else
+            addRange(dataForNode, symbol, EventOccurrencesRange.EXACTLY_ONCE, ::PathAwarePropertyInitializationInfo)
+    }
+
+    // --------------------------------------------------
+    // Data flows of declared/assigned variables in loops
+    // --------------------------------------------------
+
+    // Note that [PreliminaryLoopVisitor] in FIR DFA collects assigned variable names.
+    // This one collects declared variable symbols per capturing statements.
+    object CapturingStatementVisitor {
+        val declaredVariablesPerElement: SetMultimap<FirStatement, FirPropertySymbol> = setMultimapOf()
+
+        fun enterCapturingStatement(statement: FirStatement): Set<FirPropertySymbol> {
+            assert(statement is FirLoop || statement is FirClass<*> || statement is FirFunction<*>)
+            if (statement !in declaredVariablesPerElement) {
+                statement.accept(visitor, null)
+            }
+            return declaredVariablesPerElement[statement]
+        }
+
+        fun exitCapturingStatement(statement: FirStatement) {
+            assert(statement is FirLoop || statement is FirClass<*> || statement is FirFunction<*>)
+            declaredVariablesPerElement.removeKey(statement)
+        }
+
+        // FirStatement -- closest statement (loop/lambda/local declaration) which may contain reassignments
+        private val visitor = object : FirVisitor<Unit, FirStatement?>() {
+            override fun visitElement(element: FirElement, data: FirStatement?) {
+                element.acceptChildren(this, data)
+            }
+
+            override fun visitProperty(property: FirProperty, data: FirStatement?) {
+                if (property.isLocal) {
+                    requireNotNull(data)
+                    declaredVariablesPerElement.put(data, property.symbol)
+                }
+                visitElement(property, data)
+            }
+
+            override fun visitWhileLoop(whileLoop: FirWhileLoop, data: FirStatement?) {
+                visitCapturingStatement(whileLoop, data)
+            }
+
+            override fun visitDoWhileLoop(doWhileLoop: FirDoWhileLoop, data: FirStatement?) {
+                visitCapturingStatement(doWhileLoop, data)
+            }
+
+            override fun visitAnonymousFunction(anonymousFunction: FirAnonymousFunction, data: FirStatement?) {
+                visitCapturingStatement(anonymousFunction, data)
+            }
+
+            override fun visitSimpleFunction(simpleFunction: FirSimpleFunction, data: FirStatement?) {
+                visitCapturingStatement(simpleFunction, data)
+            }
+
+            override fun <F : FirFunction<F>> visitFunction(function: FirFunction<F>, data: FirStatement?) {
```

Comment:
You don't need to override `visitFuinction` after you override `visitXXX` for all `FirFunction<*>` inheritors

Reply:
Since lambda and local declarations are targets to visit, I think overriding `visitAnonymousFunction` and `visitSimpleFunction` is good enough (and that's what/why `PreliminaryLoopVisitor` did so). This general one, though, is added while mimicking `PreliminaryLoopVisitor`, so I guess it's safe to remove both ones? :) Added as the 2nd commit.